src/script/microbit-interfacing/MicrobitSerial.ts

```diff
@@ -4,109 +4,236 @@
+      if (value) {
+        this.unprocessedInput += value;
+
+        console.log("full", this.unprocessedInput)
+
+        let messages = protocol.splitMessages(this.unprocessedInput);
+        this.unprocessedInput = messages.remainingInput;
+
+        console.log("extra", this.unprocessedInput)
+
+        messages.messages.forEach(msg => {
+          const sensorData = protocol.processPeriodicMessage(msg);
+          if (sensorData) {
+            // FIXME: Debug log, printing the processed data and time between messages.
+            let now = Date.now();
+            console.log(now - this.t, sensorData);
+            this.t = now;
+
+            inputBehaviour.accelerometerChange(
+              sensorData.accelerometerX,
+              sensorData.accelerometerY,
+              sensorData.accelerometerZ,
+            );
+            if (sensorData.buttonA !== previousButtonState.A) {
+              previousButtonState.A = sensorData.buttonA;
+              inputBehaviour.buttonChange(sensorData.buttonA, 'A');
+            }
+            if (sensorData.buttonB !== previousButtonState.B) {
+              previousButtonState.B = sensorData.buttonB;
+              inputBehaviour.buttonChange(sensorData.buttonB, 'B');
+            }
           }
-          if (done) {
-            writeLine('Done');
-            reading = false;
-          }
-          // console.log(microbitState);
-        }
-        // clearInterval(interval);
-        reader.releaseLock();
+        });
+      }
+      // TODO: Is this necessary? Will the DAPJs read() function ever return done=true?
+      if (done) {
+        break;
       }
-    } catch (error) {
-      console.log('There was an error');
-      console.log(error);
-    } finally {
-      await serialPort.close();
     }
   }
+
+  public listenForDisconnect(callback: (event: Event) => unknown): void {}
+
+  public removeDisconnectListener(callback: (event: Event) => unknown): void {}
+
+  private async connect(): Promise<void> {
+    let serialPort = this.usb.serialPort;
+    await serialPort.open({ baudRate: this.baudRate });
+    writeLine(`Opened with baudRate: ${this.baudRate}`);
+    if (serialPort.readable && serialPort.writable) {
+      this.connected = true;
+      this.encoder.readable.pipeTo(serialPort.writable);
+      serialPort.readable.pipeTo(this.decoder.writable);
+    } else {
+      throw new Error('Serial port not readable or writable');
+    }
+  }
+
+  public isConnected(): boolean {
+    return this.connected;
+  }
+
+  public disconnect(): void {
+    this.connected = false;
+    this.usb.disconnect();
+    this.onDisconnect(true);
+    this.unprocessedInput = '';
+
+    if (this.periodicDataPromise) {
+      // It's hard to make disconnect() async so we've left this as a background error for now.
+      this.periodicDataPromise.catch(e => {
+        console.error(e);
+      });
+      this.periodicDataPromise = undefined;
+    }
+
+    navigate(Paths.HOME);
```

Comment:
This shouldn't be here, but I'm not sure where it will need to go.

Reply:
I think we should be able to remove this now. The equivalent bluetooth code was removed by Rob in 9ff240a75d0ae59bff93c7cb1c447be7d197195d.

Will test.

Reply:
In my testing we don't currently notice disconnection in radio bridge mode. I think we can remove the navigate call and merge with that as a known defect.

Reply:
22ca71c33efde4ef41b05ae74fddd4c937fd9b99

Reply:
> In my testing we don't currently notice (physical) disconnection in radio bridge mode. I think we can remove the navigate call and merge with that as a known defect.

Yes, `removeDisconnectListener()` still needs to be implemented, if a listener is register I assume that's the function calling `navigate(Paths.HOME)`?